tools/src/tensil/tools/emulator/Emulator.scala

```diff
@@ -0,0 +1,595 @@
+    writeDRAM0(addresses, arrayToStream(bytes))
+
+  def readDRAM0(addresses: Seq[MemoryAddressRaw], stream: OutputStream): Unit =
+    executive.readDRAM0(addresses, new DataOutputStream(stream))
+  def readDRAM0(addresses: Seq[MemoryAddressRaw]): Array[Byte] = {
+    val bytesStream = new ByteArrayOutputStream()
+    executive.readDRAM0(addresses, new DataOutputStream(bytesStream))
+    bytesStream.toByteArray()
+  }
+
+  def writeDRAM1(size: MemoryAddressRaw, stream: InputStream) =
+    executive.writeDRAM1(0L until size, new DataInputStream(stream))
+  def writeDRAM1(size: MemoryAddressRaw, bytes: Array[Byte]): Unit =
+    writeDRAM1(size, arrayToStream(bytes))
+
+  def run(programBytes: Array[Byte], trace: ExecutiveTrace): Unit =
+    run(new ByteArrayInputStream(programBytes), trace)
+
+  def run(stream: InputStream, trace: ExecutiveTrace): Unit = {
+    val startTime = System.nanoTime()
+
+    val traceLir = new LIR {
+
+      /**
+        * This will run the trace for inital tracepoints.
+        * See MemoryManager.emitInitialTracepoints.
+        */
+      trace.runTrace(-InstructionAddress.One, executive)
+
+      private def runTrace(context: Option[InstructionContext]): Unit =
+        trace.runTrace(context.get.address.get, executive)
+
+      def emitWait(
+          tidToWait: Int,
+          tid: Int,
+          context: Option[InstructionContext]
+      ): Unit =
+        runTrace(context)
+
+      def emitMatMul(
+          accumulate: Boolean,
+          localStride: Int,
+          localAddress: MemoryAddress,
+          accumulatorStride: Int,
+          accumulatorAddress: MemoryAddress,
+          size: MemoryAddressRaw,
+          tid: Int,
+          context: Option[InstructionContext]
+      ): Unit = runTrace(context)
+
+      def emitSIMD(
+          accumulate: Boolean,
+          simdOp: Int,
+          simdSourceLeft: Int,
+          simdSourceRight: Int,
+          simdDestination: Int,
+          writeAccumulatorAddress: MemoryAddress,
+          readAccumulatorAddress: MemoryAddress,
+          tid: Int,
+          context: Option[InstructionContext]
+      ): Unit = runTrace(context)
+
+      def emitDataMove(
+          toLocal: Boolean,
+          accumulate: Boolean,
+          localStride: Int,
+          localAddress: MemoryAddress,
+          stride: Int,
+          address: MemoryAddress,
+          size: MemoryAddressRaw,
+          tid: Int,
+          context: Option[InstructionContext]
+      ): Unit = runTrace(context)
+
+      def emitLoadWeights(
+          localStride: Int,
+          localAddress: MemoryAddress,
+          size: MemoryAddressRaw,
+          tid: Int,
+          context: Option[InstructionContext]
+      ): Unit = runTrace(context)
+
+      def endEmit(): Unit = {}
+    }
+
+    val parser =
+      lir.Parser.injectInstructionAddress(new lir.StreamParser(arch, stream))
+
+    val sequencerLir = new lir.Sequencer(
+      arch = arch,
+      readLir = executive.readLir,
```

Comment:
wow nice, using the LIR interface to implement the emulator too - this is really elegant and makes a lot of sense